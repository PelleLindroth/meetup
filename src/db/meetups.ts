import { User, users } from './users'

export interface Meetup {
  id: string
  title: string
  description?: string
  date: Date
  location?: Address
  url?: string
  online: boolean
  capacity: number | null
  attending: number
  arranger: User
  reviews: Review[]
  comments: Comment[]
}

interface Address {
  street: string
  city: string
  country: string
}

interface Review {
  id: string
  meetupId: string
  rating: number
  submittedAt: Date
}

interface Comment {
  id: string
  meetupId: string
  title: string
  body: string
  submittedAt: Date
}

export const meetups: Meetup[] = [
  {
    id: '1',
    title: 'Pizza Picnic',
    description: 'We meet in the park with pizzas and hang out all night',
    date: new Date(1630170000000),
    location: {
      street: 'Vitabergsparken',
      city: 'Stockholm',
      country: 'Sweden',
    },
    online: false,
    url: 'https://www.udemy.com/course/rust-lang/',
    capacity: null,
    attending: 54,
    arranger: users[0],
    reviews: [],
    comments: [],
  },
  {
    id: '2',
    title: 'Rust course',
    description: 'Learn a new programming language from the ground up',
    date: new Date(1641808800000),
    online: true,
    url: 'https://www.udemy.com/course/rust-lang/',
    capacity: null,
    attending: 234,
    arranger: users[2],
    reviews: [],
    comments: [],
  },
  {
    id: '3',
    title: 'Sevilla - Real Sociedad',
    description: 'A classic football match between the south and the north',
    date: new Date(1652457600000),
    location: {
      street: 'C. Sevilla Fútbol Club',
      city: '41005 Sevilla',
      country: 'Spain',
    },
    online: false,
    capacity: 43883,
    attending: 23478,
    arranger: users[1],
    reviews: [],
    comments: [],
  },
  {
    id: '4',
    title: 'Kite festival',
    description:
      'Classic Kite festival at Gärdet, bring your own kite or just watch the pros',
    date: new Date(1656774000000),
    location: {
      street: 'Gärdet',
      city: 'Stockholm',
      country: 'Sweden',
    },
    online: false,
    capacity: null,
    attending: 126,
    arranger: users[3],
    reviews: [],
    comments: [],
  },
]
