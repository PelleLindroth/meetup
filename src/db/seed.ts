import { UserBank, User } from './models/User'

const user1 = new User(
  'Tjalle',
  'Bergkvist',
  'tjalle@yahoo.com',
  'grillkorv',
  '1'
)
user1.addAttending('1')
user1.addAttending('2')
user1.addReviewed('1')

const user2 = new User('Kenta', 'Andersson', 'kenta@yahoo.com', 'bananpaj', '2')
user2.addAttending('1')
user2.addReviewed('1')

const user3 = new User(
  'Barbro',
  'Lundborg',
  'babsan@yahoo.com',
  'password123',
  '3'
)

const user4 = new User('Janne', 'Karlsson', 'loffe@yahoo.com', 'storstark', '4')
user4.addAttending('1')

const users = new UserBank()

users.add(user1)
users.add(user2)
users.add(user3)
users.add(user4)

export { users }
