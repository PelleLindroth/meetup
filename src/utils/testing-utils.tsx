import { render } from '@testing-library/react'
import { ReactElement } from 'react'
import { MemoryRouter } from 'react-router-dom'
import { Routes, Route } from 'react-router'

export const renderWithRouter = (ui: ReactElement, options?: any) =>
  render(ui, { wrapper: MemoryRouter, ...options })

export const renderWithPath = (ui: ReactElement, entry: string, path: string) =>
  render(
    <MemoryRouter initialEntries={[entry]}>
      <Routes>
        <Route path={path} element={ui} />
      </Routes>
    </MemoryRouter>
  )
