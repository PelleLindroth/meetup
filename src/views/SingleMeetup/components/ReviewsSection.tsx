import { ReviewsSectionProps } from '../types'
import styles from '../SingleMeetup.module.scss'
import StarIcon from '../../../assets/icons/star-empty.png'

const ReviewsSection = (props: ReviewsSectionProps) => {
  const { user, meetup } = props

  return (
    <section className={styles.reviews}>
      <div className={styles.reviewRow}>
        <img src={StarIcon} alt="Star icon" />
        {meetup.reviews.length ? (
          <p>{`This event has a ${
            meetup.reviews.reduce((acc, curr) => acc + curr.rating, 0) /
            meetup.reviews.length
          } score out of 5 possible after ${meetup.reviews.length} votes`}</p>
        ) : (
          <p>This event has no ratings yet</p>
        )}
      </div>
      {user &&
        user.attended.includes(meetup.id) &&
        (!user.reviewed.includes(meetup.id) ? (
          <button>Rate this event</button>
        ) : (
          <p>You already rated this event</p>
        ))}
    </section>
  )
}

export default ReviewsSection
