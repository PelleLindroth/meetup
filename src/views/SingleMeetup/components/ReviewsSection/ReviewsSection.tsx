import { useEffect, useState } from 'react'
import { ReviewsSectionProps } from '../../types'
import StarIcon from '../../../../assets/icons/star-empty.png'
import RatingForm from './components/RatingForm'
import styles from '../../SingleMeetup.module.scss'

const ReviewsSection = (props: ReviewsSectionProps) => {
  const { user, meetup } = props
  const [showSelectInput, setShowSelectInput] = useState<boolean>(false)
  const [medianRating, setMedianRating] = useState<number>(0)

  useEffect(() => {
    setMedianRating(
      meetup.reviews.reduce((acc, curr) => acc + curr.rating, 0) /
        meetup.reviews.length
    )
  }, [meetup.reviews])

  return (
    <section className={styles.reviews}>
      <div className={styles.headerRow}>
        <img src={StarIcon} alt="Star icon" />
        <h3>Ratings</h3>
      </div>
      {meetup.reviews.length ? (
        <p>
          {`This event has a `}
          <span className={styles.rating}>{`${(
            Math.round(medianRating * 100) / 100
          ).toFixed(1)}`}</span>
          {` score out of 5 possible after ${meetup.reviews.length} votes`}
        </p>
      ) : (
        <p>This event has no ratings yet</p>
      )}
      {user &&
        user.attended.includes(meetup.id) &&
        (!user.reviewed.includes(meetup.id) ? (
          <>
            {showSelectInput ? (
              <RatingForm
                meetup={meetup}
                user={user}
                setMedianRating={setMedianRating}
                setShowSelectInput={setShowSelectInput}
              />
            ) : (
              <button onClick={() => setShowSelectInput(true)}>
                Rate this event
              </button>
            )}
          </>
        ) : (
          <p>You already rated this event</p>
        ))}
    </section>
  )
}

export default ReviewsSection
