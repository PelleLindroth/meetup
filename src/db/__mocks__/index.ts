import { meetups, Meetup } from '../meetups'
import { users, User } from '../users'
import { sortMeetupsChronologically } from '../dbutils'

export const getAllMeetups = () => {
  sortMeetupsChronologically(meetups)

  return meetups
}

export const getMeetupById = (id: string): Meetup | undefined => {
  return meetups.find((meetup) => meetup.id === id)
}

export const addComment = jest.fn()

export const addReview = jest.fn()

export const addMeetup = (meetup: Meetup) => {
  meetups.push(meetup)
}

export const validateUser = (email: string, password: string): User | null => {
  const user = users.find((user) => user.email === email)

  if (user) {
    return user.password === password ? user : null
  } else {
    return null
  }
}

export const signUpForEvent = (meetup: Meetup, user: User) => {
  user.attending.push(meetup.id)
  meetup.attending++
}

export const cancelSignUpForEvent = (meetup: Meetup, user: User) => {
  user.attending = user.attending.filter((item) => item !== meetup.id)
  meetup.attending--
}

export const storeUser = jest.fn()

export const getStoredUser = jest.fn()

export const clearStoredUser = jest.fn()

export const getUserById = (id: string): User | undefined => {
  return users.find((user) => user.id === id)
}

export const getUsers = () => {
  return users
}

export const storeUserAttending = jest.fn()

export const getUserAttending = (userId: string): string[] => {
  return getUserById(userId)!.attending
}

export const storeUserReviewed = jest.fn()

export const getUserReviewed = (userId: string): string[] => {
  return getUserById(userId)!.reviewed
}
