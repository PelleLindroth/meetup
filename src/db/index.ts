import { meetups, Meetup } from './meetups'
import { users, User } from './users'

export const getAllMeetups = (): Meetup[] => {
  const storedMeetups: Meetup[] = JSON.parse(localStorage.getItem('meetups')!)
  let returnedMeetups: Meetup[] = []

  if (storedMeetups) {
    returnedMeetups.push(...storedMeetups)
  } else returnedMeetups.push(...meetups)

  returnedMeetups.sort((a, b) => {
    if (a.date.getTime() > b.date.getTime()) {
      return 1
    } else if (b.date.getTime() > a.date.getTime()) {
      return -1
    } else {
      return 0
    }
  })

  return returnedMeetups
}

export const getMockMeetups = () => {
  meetups.sort((a, b) => {
    if (a.date.getTime() > b.date.getTime()) {
      return 1
    } else if (b.date.getTime() > a.date.getTime()) {
      return -1
    } else {
      return 0
    }
  })

  return meetups
}

export const saveMeetupToLocalStorage = (meetup: Meetup) => {
  const storedMeetups: Meetup[] = JSON.parse(localStorage.getItem('meetups')!)

  if (storedMeetups) {
    storedMeetups.push(meetup)

    localStorage.setItem('meetups', JSON.stringify(storedMeetups))
  } else {
    localStorage.setItem('meetups', JSON.stringify([meetup]))
  }
}

export const getMeetupById = (id: string): Meetup | undefined => {
  return meetups.find((meetup) => meetup.id === id)
}

export const validateUser = (email: string, password: string): User | null => {
  const user = users.find((user) => user.email === email)

  if (user) {
    return user.password === password ? user : null
  } else {
    return null
  }
}

export const getUserById = (id: string): User | undefined => {
  return users.find((user) => user.id === id)
}

export const getMockUsers = () => {
  return users
}
