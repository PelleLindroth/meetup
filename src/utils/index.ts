import { User } from '../db/users'
import { Meetup } from '../db/meetups'

export const formatDate = (date: Date) => {
  return `${new Intl.DateTimeFormat('en-GB', {
    year: 'numeric',
    month: 'long',
    day: '2-digit',
    hour: 'numeric',
    minute: 'numeric',
  }).format(date)}`
}

const sortChronologically = (array: Meetup[]) => {
  return array.sort((a, b) => a.date.getTime() - b.date.getTime())
}

export const createMeetupList = (
  meetups: Meetup[],
  listType: string,
  searchPhrase: string,
  searchFilter: string
) => {
  const meetupList = meetups.filter((meetup) => {
    if (listType === 'past') {
      return (
        meetup.date.getTime() < Date.now() &&
        meetup.title.toLowerCase().includes(searchPhrase.toLowerCase()) &&
        (searchFilter === listType || searchFilter === 'all')
      )
    } else {
      return (
        meetup.date.getTime() > Date.now() &&
        meetup.title.toLowerCase().includes(searchPhrase.toLowerCase()) &&
        (searchFilter === listType || searchFilter === 'all')
      )
    }
  })

  return sortChronologically(meetupList)
}

export const createUserLists = (user: User, meetups: Meetup[]) => {
  return {
    ownUpcoming: sortChronologically(
      meetups.filter(
        (meetup: Meetup) =>
          meetup.arranger.id === user!.id &&
          meetup.date.getTime() > new Date().getTime()
      )
    ),
    ownPast: sortChronologically(
      meetups.filter(
        (meetup: Meetup) =>
          meetup.arranger.id === user!.id &&
          meetup.date.getTime() < new Date().getTime()
      )
    ),
    upcomingAttending: sortChronologically(
      meetups.filter(
        (meetup: Meetup) =>
          user!.attending.includes(meetup.id) &&
          meetup.date.getTime() > new Date().getTime()
      )
    ),
    pastAttended: sortChronologically(
      meetups.filter(
        (meetup: Meetup) =>
          user!.attending.includes(meetup.id) &&
          meetup.date.getTime() < new Date().getTime()
      )
    ),
  }
}
