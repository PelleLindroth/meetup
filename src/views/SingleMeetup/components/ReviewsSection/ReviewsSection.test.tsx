import { renderWithPath, mountWithPath } from '../../../../utils/testing-utils'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { getMeetupById, getUserById } from '../../../../db'
import ReviewsSection from './index'

describe('ReviewsSection unit tests', () => {
  const kiteEvent = getMeetupById('4')
  const halloweenEvent = getMeetupById('5')
  const pizzaEvent = getMeetupById('1')
  const userLoffe = getUserById('4')

  it('renders ReviewsSection correctly (smoke test)', () => {
    const wrapper = mountWithPath(
      <ReviewsSection meetup={pizzaEvent!} user={null} />,
      `/meetup/${pizzaEvent!.id}`,
      'meetup/:id'
    )

    expect(wrapper.find(ReviewsSection)).toMatchSnapshot()
  })
  it('renders a median rating for past event', () => {
    renderWithPath(
      <ReviewsSection meetup={pizzaEvent!} user={null} />,
      `/meetup/${pizzaEvent!.id}`,
      'meetup/:id'
    )

    const score = screen.getByText(/score/i)

    expect(score).toHaveTextContent(
      'This event has a 4.5 score out of 5 possible after 2 votes'
    )
  })
  it('renders a message if no reviews were submitted', () => {
    renderWithPath(
      <ReviewsSection meetup={halloweenEvent!} user={null} />,
      `/meetup/${halloweenEvent!.id}`,
      'meetup/:id'
    )

    const noRatingsMessage = screen.getByText(/no ratings/i)

    expect(noRatingsMessage).toHaveTextContent('This event has no ratings yet')
  })
  it('does not render a review score for upcoming event', () => {
    renderWithPath(
      <ReviewsSection meetup={kiteEvent!} user={null} />,
      `/meetup/${kiteEvent!.id}`,
      'meetup/:id'
    )

    const score = screen.queryByText(/score/i)

    expect(score).not.toBeInTheDocument()
  })
  it('renders a Rate this event button if user is logged in and has attended', () => {
    renderWithPath(
      <ReviewsSection meetup={pizzaEvent!} user={userLoffe!} />,
      `/meetup/${pizzaEvent!.id}`,
      'meetup/:id'
    )

    const rateButton = screen.getByRole('button')

    expect(rateButton).toBeInTheDocument()
  })
  it('renders a review form with select input when add review button is clicked', () => {
    renderWithPath(
      <ReviewsSection meetup={pizzaEvent!} user={userLoffe!} />,
      `/meetup/${pizzaEvent!.id}`,
      'meetup/:id'
    )

    const rateButton = screen.getByRole('button')

    userEvent.click(rateButton)

    const reviewSelectInput = screen.getByRole('combobox')

    expect(reviewSelectInput).toBeInTheDocument()
  })
  it('updates after submitted review to display new median value and number of submitted ratings', () => {
    renderWithPath(
      <ReviewsSection meetup={pizzaEvent!} user={userLoffe!} />,
      `/meetup/${pizzaEvent!.id}`,
      'meetup/:id'
    )

    const rateButton = screen.getByRole('button', { name: /rate/i })

    userEvent.click(rateButton)

    const reviewSelectInput = screen.getByRole('combobox')

    userEvent.selectOptions(reviewSelectInput, '1 - Awful')

    const submitButton = screen.getByRole('button', { name: /submit/i })

    userEvent.click(submitButton)

    const score = screen.getByText(/score/i)

    expect(score).toHaveTextContent(
      'This event has a 3.3 score out of 5 possible after 3 votes'
    )
  })
})
