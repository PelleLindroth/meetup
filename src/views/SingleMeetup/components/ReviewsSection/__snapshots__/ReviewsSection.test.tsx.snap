// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewsSection unit tests renders ReviewsSection correctly (smoke test) 1`] = `
<ReviewsSection
  meetup={
    Object {
      "arranger": UserImpl {
        "addAttending": [Function],
        "addReviewed": [Function],
        "attending": Array [
          "1",
          "2",
        ],
        "cancelAttending": [Function],
        "email": "tjalle@yahoo.com",
        "firstName": "Tjalle",
        "getDetails": [Function],
        "getFullName": [Function],
        "id": "1",
        "lastName": "Bergkvist",
        "password": "grillkorv",
        "reviewed": Array [
          "1",
        ],
        "saveDetailsToLocalStorage": [Function],
        "saveIdToLocalStorage": [Function],
      },
      "attending": 54,
      "capacity": null,
      "comments": Array [
        Object {
          "id": "1",
          "submittedAt": 2021-12-16T14:58:33.837Z,
          "text": "Great day!",
          "userId": "2",
        },
        Object {
          "id": "2",
          "submittedAt": 2021-12-16T14:59:22.526Z,
          "text": "Yes, it was, thanks!",
          "userId": "1",
        },
        Object {
          "id": "3",
          "submittedAt": 2021-12-17T12:34:42.949Z,
          "text": "Damn, I missed this",
          "userId": null,
        },
      ],
      "date": 2021-08-28T17:00:00.000Z,
      "description": "We meet in the park with pizzas and hang out all night",
      "id": "1",
      "location": Object {
        "city": "Stockholm",
        "street": "Vitabergsparken",
      },
      "online": false,
      "reviews": Array [
        Object {
          "id": "1",
          "meetupId": "1",
          "rating": 4,
        },
        Object {
          "id": "2",
          "meetupId": "1",
          "rating": 5,
        },
      ],
      "title": "Pizza Picnic",
    }
  }
  user={null}
>
  <section
    className="reviews"
  >
    <div
      className="headerRow"
    >
      <img
        alt="Star icon"
        src="star-empty.png"
      />
      <h3>
        Ratings
      </h3>
    </div>
    <p>
      This event has a 
      <span
        className="rating"
      >
        4.5
      </span>
       score out of 5 possible after 2 votes
    </p>
  </section>
</ReviewsSection>
`;
